Consumer/ConsumerManager.php

```diff
@@ -0,0 +1,210 @@
+    {
+        $this->consumeStart();
+
+        $this->newTopicToConsume = '';
+        $this->queue = $this->consumer->newQueue();
+
+        if (count($topics) > 0) {
+            array_walk($topics, $this->setQueueForEveryDefinedTopic());
+        } else {
+            array_walk($this->topics, $this->setQueueForEveryTopic());
+        }
+
+        if ($this->eventDispatcher) {
+            $this->notifyEvent(self::ORIGIN);
+        }
+
+        return $this->getMessages();
+    }
+
+    /**
+     * @return array
+     */
+    public function getTopicsConsumptionState() : array
+    {
+        return $this->topicsConsumptionState;
+    }
+
+    /**
+     * @return void
+     */
+    protected function consumeStart()
+    {
+        $metadata = $this->consumer->getMetadata(true, null, 1000);
+        if (!is_null($metadata)) {
+            foreach ($metadata->getTopics() as $topic) {
+                foreach ($topic->getPartitions() as $partition) {
+                    if (!is_null($partition->getId())) {
+                        $this->topicsFromServer[$topic->getTopic()][] = $partition->getId();
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * @return \Closure
+     */
+    protected function setQueueForEveryDefinedTopic() : callable
+    {
+        return function ($partition, $topicName) {
+            $this->newTopicToConsume = $this->consumer->newTopic($topicName);
+            array_walk($partition, $this->startConsumingEveryPartition());
+        };
+    }
+
+    /**
+     * @return \Closure
+     */
+    protected function startConsumingEveryPartition() : callable
+    {
+        return function ($offset, $partition) {
+            $this->newTopicToConsume->consumeQueueStart($partition, $offset, $this->queue);
+        };
+    }
+
+    /**
+     * @return \Closure
+     */
+    protected function setQueueForEveryTopic() : callable
+    {
+        return function ($topic) {
+            $topicName = $topic->getName();
+            $this->newTopicToConsume = $this->consumer->newTopic($topicName);
+            if (isset($this->topicsFromServer[$topicName])) {
+                array_walk($this->topicsFromServer[$topicName], function ($partition) {
+                    $this->newTopicToConsume->consumeQueueStart($partition, RD_KAFKA_OFFSET_BEGINNING, $this->queue);
+                });
+            }
+        };
+    }
+
+    /**
+     * @return array
+     */
+    protected function getMessages() : array
+    {
+        $stop = false;
+        $messages = [];
+        while (!$stop) {
+            $msg = $this->queue->consume(1000);
+
+            if (is_null($msg)) {
+                $stop = true;
+            }
+
+            if (empty($msg->payload)) {
+                continue;
+            }
+
+            $messages[] = $msg;
+            $this->topicsConsumptionState[$msg->topic_name][$msg->partition] = $msg->offset;
+        }
+
+        return $messages;
```

Comment:
I'm not usual with Kafka or the global process, but are we **sure** that the queue won't be filled by an other process in the same time? Else we will be stuck in an infinite loop…

What about using [generators](http://php.net/manual/en/language.generators.syntax.php) here? If we *just* need to process messages one by one, we could `yield` the message as soon as received.
❓